Introduce Tooltip for the BlockClock navbar button

johnny9 opened the pull request:
The Tooltip will appear when hovering over the BlockClock tab button in the desktop navigation bar. It will show the current state of IBD.

Two things are different than the figma. The wording for remaining blocks matches what is shown on the BlockClock instead of what the figma shows and After sync, Blockheight is shown instead. Looking for opinions on both during review.

![vlcsnap-2024-05-01-20h05m06s361](https://github.com/bitcoin-core/gui-qml/assets/985648/fb4b6da1-ac76-4e82-b515-977f09caee51)

[![Build Artifacts](https://img.shields.io/badge/Build%20Artifacts-green
)](https://github.com/bitcoin-core/gui-qml/actions/runs/9012143092?pr=400)

MarnixCroes commented on src/Makefile.qt.include:
```diff
@@ -360,6 +361,7 @@ QML_RES_QML = \
   qml/components/StorageSettings.qml \
   qml/components/ThemeSettings.qml \
   qml/components/TotalBytesIndicator.qml \
+  qml/components/Tooltip.qml \
```
forgot to add this?

johnny9 replied:
whoops

johnny9 replied:
thank you, updated the pr with the necessary files

johnny9 commented:
Update from dc31a22 to[ a7e4faa](https://github.com/bitcoin-core/gui-qml/commit/a7e4faa438e99d088d2773968613235cadb4c518)
 - Now with tooltip

MarnixCroes reviewed (commented):
7453346a1ac518250a5600e3d37c4100c55feb8b
there should also be a tooltip state for when node is offline/stopped, i.e:
I pause the blockclock -> tooltip keeps displaying the blocktime from last online/sync time.
doesn't seem to be right

johnny9 commented:
> [7453346](https://github.com/bitcoin-core/gui-qml/commit/7453346a1ac518250a5600e3d37c4100c55feb8b) there should also be a tooltip state for when node is offline/stopped, i.e: I pause the blockclock -> tooltip keeps displaying the blocktime from last online/sync time. doesn't seem to be right

Good point. I will add some basic state status for paused/connected/offline

pablomartin4btc reviewed (commented):
tACK 7453346a1ac518250a5600e3d37c4100c55feb8b

<details>
<summary>I agree with wording about time "<b>left</b>" matching the BlockClock instead of "<b>to go</b>" as in <a href="https://www.figma.com/file/ek8w3n3upbluw5UL2lGhRx/Bitcoin-Core-App-Design?type=design&node-id=8842-850575&mode=design">figma<a>, perhaps figma needs to be updated @GBKS?</summary>

- Figma:

  ![image](https://github.com/bitcoin-core/gui-qml/assets/110166421/5b08a375-064c-470e-ab6a-b381fb8ece24)

- This PR:

  ![image](https://github.com/bitcoin-core/gui-qml/assets/110166421/9c901953-9730-47e6-8f01-a73864397df9)

</details>

It would be good to add the figma link and screenshot comparison in the description so reviewers could ACK easier. 

I also agree with @johnny9 and @MarnixCroes about adding different states wording as currently one can see the some inconsistencies:

<details>
<summary><b>Downloading blocks</b> (this one looks correct)</summary>

![image](https://github.com/bitcoin-core/gui-qml/assets/110166421/f1cc8c77-5ede-491c-907b-454ad7c5b7a1)

</details>

<details>
<summary><b>Paused</b></summary>

![Screenshot from 2024-05-06 11-53-38](https://github.com/bitcoin-core/gui-qml/assets/110166421/e8b0f5c9-01e3-49ca-a68e-b2d93cb01eb7)

</details>

<details>
<summary><b>Connecting</b> (after pausing)</summary>

![image](https://github.com/bitcoin-core/gui-qml/assets/110166421/678d8c2e-03a9-4cfe-9a76-ff7c97267db1)

</details>

<details>
<summary><b>Connecting</b> (after sync completed/ blocktime determined)</summary>

![image](https://github.com/bitcoin-core/gui-qml/assets/110166421/bf97c2b4-caaa-48a6-b4c4-5ba594582c99)

</details>

<details>
<summary><b>Blocktime</b> determined (this one looks correct)</summary>

![image](https://github.com/bitcoin-core/gui-qml/assets/110166421/a0024b6d-3e47-4406-ae37-bfde7e62b3db)

</details>

johnny9 commented:
Update  from 7453346 to bd62c7c 
 - Added paused and connecting stated
 
 
![Screenshot from 2024-05-06 21-59-03](https://github.com/bitcoin-core/gui-qml/assets/985648/7771f1da-5f34-426a-9115-f979a18c730b)
![Screenshot from 2024-05-06 21-58-56](https://github.com/bitcoin-core/gui-qml/assets/985648/f34397fb-30a8-4bd3-bfe8-91385bd9f2d8)
![Screenshot from 2024-05-06 21-58-53](https://github.com/bitcoin-core/gui-qml/assets/985648/b028ece7-65aa-47eb-85af-aa9de525d661)

johnny9 commented:
Update from bd62c7c to 2478b9e 
 - newlline at end of file
 - Added copyright header to Tooltip.qml

GBKS commented:
Good one. I tested and found a few things.

First, the screenshots:

![image](https://github.com/bitcoin-core/gui-qml/assets/695901/316405fe-ef08-43b8-962d-271b2f63cc2f)

As you can see, the text disappears in light mode, and the color of the tip does not adjust. On the web, it's possible to tie SVG colors to CSS color variables that automatically change when the theme changes. Not sure what the technique is for QML.

For me, it switched straight from "Downloading blocks - Estimating" to "Blocktime - xyz". It then counted blocks from ~1.5M or so to the current number. For much of that time it showed the almost full red circle, which felt like something was broken.

The tooltip should be placed 4-5px higher up, so the arrow tip overlaps with the button. From Figma:
![image](https://github.com/bitcoin-core/gui-qml/assets/695901/f08a9f10-b45c-4ed6-be13-c058498c1033)

Good call with the extra states. Can we list out which states we need? I started putting some of them together.
![image](https://github.com/bitcoin-core/gui-qml/assets/695901/9cd3f41d-6402-442e-996a-5ebebbf3d17f)

GBKS commented:
Also a question. I see text directly in the code. To simplify localization, should those be text reference IDs instead so the actual copy can be grabbed from a locale-specific dictionary?

johnny9 commented:
> Also a question. I see text directly in the code. To simplify localization, should those be text reference IDs instead so the actual copy can be grabbed from a locale-specific dictionary?

the qsTr() wrapper function is how localization is managed with Qt

johnny9 commented:
> For me, it switched straight from "Downloading blocks - Estimating" to "Blocktime - xyz". It then counted blocks from ~1.5M or so to the current number. For much of that time it showed the almost full red circle, which felt like something was broken.

Unfortunately that is an old bug we're aware of. It has to do with how we determine if IBD is finished or not.

> Good call with the extra states. Can we list out which states we need? 

I think you got all of them. "Could not connect" is not included in my commit as we don't currently have that state until we integrate Network detection.

johnny9 commented:
Update to  to 8f8341f:
 - Fix tooltip text in lightmode
 - Adjusted the anchor upward 5 pixels so it aligns with the main component of the BlockClock mini button
 - Added light mode version of the tooltip arrow

D33r-Gee reviewed (commented):
tACK 8f8341f6598afaa44e0370a67c7379fab78dd34c on WSL Ubuntu 22.04 LGTM

<details>
<summary>Screenshot</summary>

![Screenshot 2024-05-14 110447](https://github.com/bitcoin-core/gui-qml/assets/111142327/4c4488ef-4ac2-4fc6-a195-87ae67e90bbd)


</details>

MarnixCroes reviewed (approved):
tack 8f8341f6598afaa44e0370a67c7379fab78dd34c

one q: why are `tooltip-arrow-dark.png` and `tooltip-arrow-light.png` different size?

pablomartin4btc reviewed (approved):
re-tACK 8f8341f6598afaa44e0370a67c7379fab78dd34c 

Checked all changes from 2478b9eda056f34fbdde46813f07f5141e0f2e3b as well.
